### PR TITLE
Start to clean up lint checks (fixing edge cases), and add tests.

### PR DESCRIPTION
**Pull request by ceball**
* Add tests of lint lines per cell.

* FIx lint_lines_per_cell to handle the default (-1, meaning "don't check", currently always fails). Also simplify.

* Add tests of lint cells per notebook (initially failing one case - 0).

* Fix lint cells per notebook to handle 0 (could be used to disallow code cells). Also simplify.

* Add tests of lint function definitions per notebook (initiall failing one case - 0).

* Fix lint functions definitions per notebook to handle 0 (could be used to disallow fns).

* Simplify lint function definitions per notebook.

* Add test of class definitions per notebook (initially failing one case - 0).

* Fix lint class definitions per notebook to handle 0.

* Simplify lint class definitions per notebook function.

* Add tests for lint cell coverage.

* Simplify lint cell coverage.

* Add some TODOs/notes.

### CODE REVIEW

File: jupyterlab_celltests/lint.py [hunk 1]
```diff
@@ -7,71 +7,52 @@
...
-    if lines_per_cell:
-        for i, lines_in_cell in enumerate(metadata.get('cell_lines', [])):
-            if lines_in_cell <= lines_per_cell:
-                ret.append(LintMessage(i+1, 'Checking lines in cell', LintType.LINES_PER_CELL, True))
-            else:
-                ret.append(LintMessage(i+1, 'Checking lines in cell', LintType.LINES_PER_CELL, False))
-                passed = False
-    return ret, passed
+    if lines_per_cell < 0:
+        return [], True
```
Comment (vidartf) [inline jupyterlab_celltests/lint.py]: This would just be a silent failure, right? I would either raise an error, or ensure all cells give a lint. Same below.

File: jupyterlab_celltests/lint.py [hunk 2]
```diff
@@ -7,71 +7,52 @@
...
-            ret.append(LintMessage(-1, 'Checking classes per notebook', LintType.CLASS_DEFINITIONS, False))
-            passed = False
-    return ret, passed
+    if class_definitions < 0:
+        return [], True
+    passed = metadata.get('classes', -1) <= class_definitions
+    return [LintMessage(-1, 'Checking classes per notebook', LintType.FUNCTION_DEFINITIONS, passed)], passed
 
 
+# TODO: I think this isn't lint and should be removed.
```
Comment (vidartf) [inline jupyterlab_celltests/lint.py]: I'd argue this is indeed a lint (but maybe of a different kind).

File: jupyterlab_celltests/lint.py [hunk 3]
```diff
@@ -7,71 +7,52 @@
...
-    if cell_coverage:
-        if (metadata.get('test_count', 0)/metadata.get('cell_count', -1))*100 >= cell_coverage:
-            ret.append(LintMessage(-1, 'Checking cell test coverage', LintType.CELL_COVERAGE, True))
-        else:
-            ret.append(LintMessage(-1, 'Checking cell test coverage', LintType.CELL_COVERAGE, False))
-            passed = False
-    return ret, passed
+    if cell_coverage < 0:
+        return [], True
+    passed = 100*metadata.get('test_count', 0)/metadata.get('cell_count', -1) >= cell_coverage
```
Comment (vidartf) [inline jupyterlab_celltests/lint.py]: Maybe add a docstring that coverage is expected as percent (0-100) instead of normalized (0-1).

File: jupyterlab_celltests/lint.py (hunk 2 above)
Comment (ceball) [reply jupyterlab_celltests/lint.py]: Isn't it just "test coverage", which people would usually expect to see named as such (rather than being under lint)?

File: jupyterlab_celltests/lint.py (hunk 1 above)
Comment (ceball) [reply jupyterlab_celltests/lint.py]: `lines_per_cell` is the specification of max lines allowed per cell. The existing convention in celltests (unrelated to this PR) is that -1 means "don't check lines per cell".

Before this PR, `lines_per_cell` of -1 would (presumably accidentally) result in a lint failure for every cell, because every cell would have > -1 lines in it.

Now, `lines_per_cell` of -1 passes no matter how many lines per cell ("do not check the lines per cell"). Whether -1 is the right way to achieve that is a different question, but I don't think it's related to this PR (as it was already that way).

I believe a similar thing applies to other cases below. 

So, for instance, previously if `class_definitions` (max class definitions per notebook) was set to 0 - which ought to mean no class definitions allowed - then any number of class definitions was permitted (no lint failure). If `class_definitions` was set to -1, any number of actual class definitions (including 0) would fail.

File: jupyterlab_celltests/lint.py (hunk 3 above)
Comment (ceball) [reply jupyterlab_celltests/lint.py]: Will do, good idea (though there's no change in this PR to that aspect).

File: jupyterlab_celltests/lint.py (hunk 1 above)
Comment (ceball) [reply jupyterlab_celltests/lint.py]: You're saying you'd like to see a 'lines per cell lint passed' for every cell, even if the lines per cell lint is not actually being checked? (Whereas in the PR as it is, I have gone for returning an empty list to indicate nothing was checked. I believe/guess that was the intent of the original code, which was returning an empty list if the `if lines_per_cell` test was falsey.)
